tests/conftest.py

```diff
@@ -62,18 +75,405 @@ def conn_params(test_session):
+        make_ensemble_dfvs(ensemble_1, dfv_dsg_time_series_21),
+    ]
+
+
+# TODO: Make this, or something like it, drive ensemble_dfvs_1
+@pytest.fixture(scope='function')
+def ensemble_1_data_files(data_file_1, data_file_2):
+    return {data_file_1, data_file_2}
+
+
+@pytest.fixture(scope='function')
+def ensemble_dfvs_2(ensemble_2, dfv_dsg_time_series_21, dfv_dsg_time_series_31):
+    return [
+        make_ensemble_dfvs(ensemble_2, dfv_dsg_time_series_21),
+        make_ensemble_dfvs(ensemble_2, dfv_dsg_time_series_31),
+    ]
+
+
+# TODO: Make this, or something like it, drive ensemble_dfvs_2
+@pytest.fixture(scope='function')
+def ensemble_2_data_files(data_file_3, data_file_2):
+    return {data_file_2, data_file_3}
+
+
+# Database sessions
+
+@pytest.fixture(scope="function")
+def mm_test_session_objects(
+    model_1,
+    emission_1,
+    emission_2,
+    run_11,
+    run_12,
+    data_file_1,
+    data_file_2,
+    data_file_3,
+    dfv_dsg_time_series_11,
+    dfv_dsg_time_series_12,
+    dfv_dsg_time_series_21,
+    dfv_dsg_time_series_31,
+    ensemble_1,
+    ensemble_2,
+    ensemble_dfvs_1,
+    ensemble_dfvs_2,
+):
+    # Note: Order matters. These objects must be inserted in the order given,
+    # and removed in the reverse order.
+    # Note: Completeness matters. SQLAlchemy will implicitly add objects but
+    # not necessarily implicitly delete them.
+    return (
+        [
+            model_1,
+            emission_1,
+            emission_2,
+            run_11,
+            run_12,
+            data_file_1,
+            data_file_2,
+            data_file_3,
+            dfv_dsg_time_series_11,
+            dfv_dsg_time_series_12,
+            dfv_dsg_time_series_21,
+            dfv_dsg_time_series_31,
+            ensemble_1,
+            ensemble_2,
+        ] +
+        ensemble_dfvs_1 +
+        ensemble_dfvs_2
+    )
+
+
 @pytest.fixture(scope="function")
-def ensemble_member_lister():
-    return EnsembleMemberLister(modelmeta.test_dsn)
+def mm_test_session(mm_empty_session, mm_test_session_objects):
+    s = mm_empty_session
+    for obj in mm_test_session_objects:
+        s.add(obj)
+        s.flush()
+    yield s
+
 
+# TODO: Consider substituting delete actions for rollback everywhere
 @pytest.fixture(scope="function")
-def raster_metadata():
-    return RasterMetadata(modelmeta.test_dsn)
+def mm_test_session_committed(mm_test_session, mm_test_session_objects):
+    # Contents of an uncommitted session can only be seen by that session;
+    # i.e., a session is implicitly in a transaction. This is good.
+    # Some components of pdp_util, e.g., EnsembleCatalog, creates an
+    # independent engine and session to access the database, so we must commit
+    # our session contents.
+    # However, committing a session leaves gunk in the database that can
+    # mess up the database setup for other tests. Hence we have to clean up
+    # after ourselves. And commit that cleanup.
+
```

Comment:
This seems like a reasonable approach to take. Though, I wonder whether anything can be done with multi-level transactions or savepoints to accomplish the same thing?

Reply:
Yes; this was faster than looking up how to do that and making it work.

Reply:
Ultimately, no. As I have noted in new comments:

> Some of the subjects under test establish a completely independent
connection with the database, taking only the database DSN (not a session) as
an argument. Therefore test contents in the database for such subjects must be
fully and irrevocably committed to the database. Such a commit *cannot* be
surrounded by a transaction that is later used to roll back the commit; the
objects thus committed are only visible within the transaction, and that
necessarily excludes independent database connections. Therefore we have
fixtures that truly commit objects to the database. To undo such commits they
must delete those objects.